shared/stateutil/arrays.go

```diff
@@ -0,0 +1,158 @@
+package stateutil
+
+import (
+	"bytes"
+	"errors"
+	"sync"
+
+	"github.com/minio/highwayhash"
+	"github.com/protolambda/zssz/merkle"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+)
+
+var (
+	fastSumHashKey = bytesutil.ToBytes32([]byte("hash_fast_sum64_key"))
+	leavesCache    = make(map[string][][]byte)
+	layersCache    = make(map[string][][][]byte)
+	lock           sync.Mutex
+)
+
+func arraysRoot(roots [][]byte, fieldName string) ([32]byte, error) {
+	lock.Lock()
+	if _, ok := layersCache[fieldName]; !ok {
```

Comment:
Does this need to be it's own lock sequence? Would it be possible to group it with the other lock below?

Reply:
I think it's OK to have minimal locking groups.